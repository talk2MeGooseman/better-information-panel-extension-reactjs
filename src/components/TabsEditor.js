import AppBar from '@material-ui/core/AppBar';
import { withStyles } from '@material-ui/core/styles';
import Tab from '@material-ui/core/Tab';
import Tabs from '@material-ui/core/Tabs';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import React from 'react';
import SwipeableViews from 'react-swipeable-views';
import Editor from './Editor';

const styles = theme => ({
  root: {
    backgroundColor: theme.palette.background.paper,
    maxHeight: '500px',
  },
  tabRoot: {
    minWidth: theme.spacing.unit * 8,
  },
});

@observer
class TabsEditor extends React.Component {
  static propTypes = {
    tabsStore: PropTypes.object.isRequired,
    classes: PropTypes.object.isRequired,
    theme: PropTypes.object.isRequired,
    allowPreview: PropTypes.bool,
  };

  state = {
    value: 0,
  };

  // componentWillReact() {
  //   // let { tabsStore } = this.props;

  //   if (this.state.value >= tabsStore.tabCount) {
  //     this.setState({ value: tabsStore.tabCount - 1 });
  //   }
  // }

  handleChange = (event, value) => {
    let { tabsStore } = this.props;
    tabsStore.tabIndex = value;
  };

  handleChangeIndex = index => {
    let { tabsStore } = this.props;
    tabsStore.tabIndex = index;
  };

  renderTabs() {
    let { classes, tabsStore } = this.props;
    return tabsStore.tabs.map(tab => {
      return (
        <Tab
          key={tab.id}
          label={tab.title}
          classes={{ root: classes.tabRoot }}
        />
      );
    });
  }

  renderEditor() {
    let { tabsStore, allowPreview } = this.props;

    return tabsStore.tabs.map(tab => (
      <Editor
        key={tab.id}
        tab={tab}
        tabsStore={tabsStore}
        allowPreview={allowPreview}
      />
    ));
  }

  render() {
    const { classes, theme, tabsStore } = this.props;

    return (
      <div className={classes.root}>
        <AppBar position="static" color="default">
          <Tabs
            value={tabsStore.tabIndex}
            onChange={this.handleChange}
            indicatorColor="primary"
            textColor="primary"
            scrollButtons="auto"
            variant="scrollable"
          >
            {this.renderTabs()}
          </Tabs>
        </AppBar>
        <SwipeableViews
          axis={theme.direction === 'rtl' ? 'x-reverse' : 'x'}
          index={tabsStore.tabIndex}
          enableMouseEvents={true}
          onChangeIndex={this.handleChangeIndex}>
          {this.renderEditor()}
        </SwipeableViews>
      </div>
    );
  }
}

export default withStyles(styles, { withTheme: true })(TabsEditor);
