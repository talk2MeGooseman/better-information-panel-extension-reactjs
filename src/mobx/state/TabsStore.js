import { observable, computed, action, makeObservable } from "mobx";
import TabModel from "../model/TabModel";
import { uuid } from "../../services/Utils";
import {
  queryPanelInformation
} from "../../services/Ebs";
import {
  DEFAULT_BODY_TEXT,
  ACTIVE_STEP_1,
  ACTIVE_STEP_2,
  ACTIVE_STEP_3,
  ACTIVE_STEP_4,
} from "../../services/constants";
import { path } from 'ramda';

export default class TabsStore {
  tabs = [];
  client = null;
  saveState = "";
  loadingState = "pending";
  activeStep = ACTIVE_STEP_1;
  videoOverlayHeight = '80vh';
  videoComponentVisibility = true;
  videoComponentTransparent = true;
  videoToggleImageUrl = '';
  videoToggleButtonPosition = '';
  tabIndex = 0;

  constructor() {
    makeObservable(this, {
      tabs: observable,
      client: observable,
      saveState: observable,
      loadingState: observable,
      activeStep: observable,
      videoOverlayHeight: observable,
      videoComponentVisibility: observable,
      videoComponentTransparent: observable,
      videoToggleImageUrl: observable,
      videoToggleButtonPosition: observable,
      tabIndex: observable,
      tabCount: computed,
      fetchTabs: action,
      updateTabs: action,
      saveTabs: action
    });
  }

  get tabCount() {
    return this.tabs.length;
  }

  setVideoComponentVisability(value) {
    this.saveState = '';
    this.videoComponentVisibility = value;
  }

  setVideoComponentTransparency(value) {
    this.saveState = '';
    this.videoComponentTransparent = value;
  }

  setVideoToggleImageUrl(value) {
    this.saveState = '';
    this.videoToggleImageUrl = value;
  }

  setVideoToggleButtonPosition(value) {
    this.saveState = '';
    this.videoToggleButtonPosition = value;
  }

  addTab() {
    this.saveState = '';

    if (this.activeStep === ACTIVE_STEP_1) {
      this.activeStep = ACTIVE_STEP_2;
    }

    this.tabs.push(TabModel.fromJS(this, {
      id: uuid(),
      title: 'Add Title',
      textColor: '#000000',
      bgColor: '#ffffff',
      body: DEFAULT_BODY_TEXT,
    }));
  }

  fetchTabs() {
    this.loadingState = "pending"
    queryPanelInformation(this.client).then(
      // inline created action
      action("fetchSuccess", ({ data }) => {
        this.loadingState = "done"

        const result = path(['channel', 'bipConfig'], data)
        console.log(result)

        if (!result.tabs) {
          this.addTab();
          return;
        }

        this.activeStep = ACTIVE_STEP_4;

        // Check if the setting are defined from server, could be undefined
        // And screw shit up, use default state
        if (result.transparent !== undefined) {
          this.videoComponentTransparent = result.transparent;
        }
        if (result.visibility !== undefined) {
          this.videoComponentVisibility = result.visibility;
        }
        if (result.togglePosition !== undefined) {
          this.videoToggleButtonPosition = result.togglePosition;
        }

        this.tabs = result.tabs.map((tab) => {
          return TabModel.fromJS(this, tab);
        })
      }),
      // inline created action
      action("fetchError", error => {
        this.loadingState = "error"
      })
    )
  }

  updateTabs() {
    queryPanelInformation(this.client).then(
      // inline created action
      action("fetchSuccess", result => {
        if (!result.tabs) {
          this.addTab();
          return;
        }
        this.activeStep = ACTIVE_STEP_4;
        this.tabs = result.tabs.map((tab) => {
          return TabModel.fromJS(this, tab);
        })
      }),
      // inline created action
      action("fetchError", error => {
        this.loadingState = "error"
      })
    )
  }

  saveTabs() {
    this.saveState = "pending"
      setPanelInformation(this.token, this.toJSON()).then(
        // inline created action
        action("fetchSuccess", result => {
          this.saveState = "done"
          this.activeStep = ACTIVE_STEP_4;
        }),
        // inline created action
        action("fetchError", error => {
          this.saveState = "error"
        })
      )
  }

  toJSON() {
    let jTabs = this.tabs.map((tab) => {
      return tab.toJSON();
    });

    return {
      tabs: jTabs,
      visibility: this.videoComponentVisibility,
      transparent: this.videoComponentTransparent,
      togglePosition: this.videoToggleButtonPosition,
    };
  }
}
