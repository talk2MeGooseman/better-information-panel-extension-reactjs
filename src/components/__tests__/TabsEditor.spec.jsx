import React from "react";
import { createShallow, createMount} from '@material-ui/core/test-utils';

import TabsEditor from "../TabsEditor";
import TabsStore from "../../mobx/state/TabsStore";

describe('TabsEditor', () => {
  let tabsStore, tab, shallow, mount;

  afterEach(() => {
    mount.cleanUp();
  });

  beforeEach(() => {
    shallow = createShallow({
      untilSelector: 'Tab'
    });
    mount = createMount();
    tabsStore = new TabsStore();
    tabsStore.addTab();
    tab = tabsStore.tabs[0];
    tab.id = "1";
  });

  it('match snapshot in default state', () => {
    let wrapper = shallow(<TabsEditor tabsStore={tabsStore} />)
    expect(wrapper).toMatchSnapshot();
  });

  it("should display tab with its title and editor", () => {
    let wrapper = shallow(<TabsEditor tabsStore={tabsStore} />)
    expect(wrapper.find({ label: 'Add Title' }).exists()).toBeTruthy();
    tab.title = 'Lame';
    wrapper.update();
    expect(wrapper.find({ label: 'Lame' }).exists()).toBeTruthy();
    expect(wrapper.find('Editor').length).toEqual(1);
  });

  it("should display a tab and editor for each tab in store", () => {
    tabsStore.addTab();
    let wrapper = shallow(<TabsEditor tabsStore={tabsStore} />)
    expect(wrapper.find({ label: 'Add Title' }).length).toEqual(2);
    expect(wrapper.find('Editor').length).toEqual(2);
  });
});