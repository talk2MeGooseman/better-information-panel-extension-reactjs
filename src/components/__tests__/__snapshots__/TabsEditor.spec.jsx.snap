// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsEditor match snapshot in default state 1`] = `
<div
  className="TabsEditor-root-1"
>
  <WithStyles(AppBar)
    color="default"
    position="static"
  >
    <WithStyles(Tabs)
      indicatorColor="primary"
      onChange={[Function]}
      scrollButtons="on"
      scrollable={true}
      textColor="primary"
      value={0}
    >
      <WithStyles(Tab)
        classes={
          Object {
            "root": "TabsEditor-tabRoot-2",
          }
        }
        key="1"
        label="Add Title"
      />
    </WithStyles(Tabs)>
  </WithStyles(AppBar)>
  <ReactSwipableView
    animateHeight={false}
    animateTransitions={true}
    axis="x"
    disableLazyLoading={false}
    disabled={false}
    enableMouseEvents={false}
    hysteresis={0.6}
    ignoreNativeScroll={false}
    index={0}
    onChangeIndex={[Function]}
    resistance={false}
    springConfig={
      Object {
        "delay": "0s",
        "duration": "0.35s",
        "easeFunction": "cubic-bezier(0.15, 0.3, 0.25, 1)",
      }
    }
    threshold={5}
  >
    <Editor
      key="1"
      tab={
        TabModel {
          "bgColor": "#ffffff",
          "body": "## Features:
        * Multiple tabs in one panel
        * Easy editing and previewing
        * Customize your text color and background color
        * Enable as a panel and video overlay
        
        **Steps to success:**
        - [x] Install the extension
        - [x] Enable is as _Panel_ or _Video Overlay_
        - [ ] Enter your text
        - [ ] Give your viewers all the information they need",
          "id": "1",
          "store": TabsStore {
            "activeStep": 1,
            "loadingState": "pending",
            "saveState": "",
            "tabs": Array [
              [Circular],
            ],
            "token": undefined,
            "videoOverlayHeight": undefined,
          },
          "textColor": "#000000",
          "title": "Add Title",
        }
      }
      tabsStore={
        TabsStore {
          "activeStep": 1,
          "loadingState": "pending",
          "saveState": "",
          "tabs": Array [
            TabModel {
              "bgColor": "#ffffff",
              "body": "## Features:
        * Multiple tabs in one panel
        * Easy editing and previewing
        * Customize your text color and background color
        * Enable as a panel and video overlay
        
        **Steps to success:**
        - [x] Install the extension
        - [x] Enable is as _Panel_ or _Video Overlay_
        - [ ] Enter your text
        - [ ] Give your viewers all the information they need",
              "id": "1",
              "store": [Circular],
              "textColor": "#000000",
              "title": "Add Title",
            },
          ],
          "token": undefined,
          "videoOverlayHeight": undefined,
        }
      }
    />
  </ReactSwipableView>
</div>
`;
